### PR TITLE
rename Scalar::Bits to Scalar::Raw and bits field to data

### PR DESCRIPTION
**Pull request by RalfJung**
Also use this opportunity to seal some abstraction leaks (other modules constructing `Scalar::Bits` directly instead of using a constructor).

r? @oli-obk

### CODE REVIEW

**Comment by oli-obk** [comment]
https://github.com/rust-lang/rust/pull/59276 has priority over this (there will be conflicts) cc @varkor

**Comment by oli-obk** [comment]
r=me after rebase

**Comment by RalfJung** [comment]
I decided to also refactor how we are doing the scalar sanity checks a bit.  And I'd like to see how expensive these checks are.

@bors try

**Comment by RalfJung** [comment]
Oh and I found a bug in the signed pointer offset code that could lead to those sanity checks failing. ;)

File: src/librustc/mir/interpret/pointer.rs [hunk 1]
```diff
@@ -20,45 +20,48 @@ pub trait PointerArithmetic: layout::HasDataLayout {
...
-            val.overflowing_sub(n)
+            let res = val.overflowing_sub(n);
+            self.truncate_to_ptr(res)
         } else {
             self.overflowing_offset(val, i as u64)
         }
     }
+
+    #[inline]
+    fn offset<'tcx>(&self, val: u64, i: u64) -> EvalResult<'tcx, u64> {
```
Comment (oli-obk) [inline src/librustc/mir/interpret/pointer.rs]: why did you move these two functions?

File: src/librustc/mir/interpret/value.rs [hunk 2]
```diff
@@ -294,12 +294,35 @@ impl<'tcx, Tag> Scalar<Tag> {
...
+            Scalar::Ptr(ptr) => {
+                assert_eq!(target_size, cx.data_layout().pointer_size);
+                Err(ptr)
+            }
+        }
+    }
+
     #[inline]
     pub fn to_bits(self, target_size: Size) -> EvalResult<'tcx, u128> {
         match self {
```
Comment (oli-obk) [inline src/librustc/mir/interpret/value.rs]: deduplicate with `to_bits_or_ptr` by calling that function

Comment (RalfJung) [reply src/librustc/mir/interpret/value.rs]: I'd like to, but that needs an `impl HasDataLayout`.

File: src/librustc/mir/interpret/pointer.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc/mir/interpret/pointer.rs]: To make this sequence of functions easier to read: functions only call functions further up.

Comment (Centril) [inline src/librustc/mir/interpret/pointer.rs]:
```suggestion
    fn offset(&self, val: u64, i: u64) -> EvalResult<'_, u64> {
```

File: src/librustc/mir/interpret/pointer.rs [hunk 3]
```diff
@@ -20,45 +20,48 @@ pub trait PointerArithmetic: layout::HasDataLayout {
...
+        let (res, over) = self.overflowing_offset(val, i);
+        if over { err!(Overflow(mir::BinOp::Add)) } else { Ok(res) }
+    }
+
+    #[inline]
+    fn signed_offset<'tcx>(&self, val: u64, i: i64) -> EvalResult<'tcx, u64> {
```
Comment (Centril) [inline src/librustc/mir/interpret/pointer.rs]:
```suggestion
    fn signed_offset(&self, val: u64, i: i64) -> EvalResult<'_, u64> {
```

**Comment by RalfJung** [comment]
@rust-timer build 21894212635c3e546e712b5546bc58aa3dac3595

File: src/librustc/mir/interpret/value.rs (hunk 2 above)
Comment (oli-obk) [reply src/librustc/mir/interpret/value.rs]: hmm... maybe then just pull out `Scalar::Raw` arm into a private method?

**Comment by RalfJung** [comment]
Timing looks much better than I expected!  max <4%.  I'll still demote some of these to `debug_assert`, but that means debug-assert-enabled builds check things much more now than they used to.

**Comment by RalfJung** [comment]
Now with fewer assertions.

@bors try

File: src/librustc/mir/interpret/pointer.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc/mir/interpret/pointer.rs]: Oh, I didn't know you could use `'_` in return position. Nice :)

File: src/librustc/mir/interpret/pointer.rs (hunk 3 above)
Comment (RalfJung) [reply src/librustc/mir/interpret/pointer.rs]: Accepting suggestions seem broken though so I am going to do this by hand.

File: src/librustc/mir/interpret/pointer.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc/mir/interpret/pointer.rs]: Oh, wait, no -- that does the wrong thing. It makes the return value take the lifetime of `&self`.

File: src/librustc/mir/interpret/value.rs (hunk 2 above)
Comment (RalfJung) [reply src/librustc/mir/interpret/value.rs]: Tried that, at least for the part that can also be shared with `Debug::fmt`.

File: src/librustc/mir/interpret/value.rs [hunk 4]
```diff
@@ -134,6 +132,12 @@ impl<Tag> fmt::Display for Scalar<Tag> {
 }
 
 impl<'tcx> Scalar<()> {
+    #[inline]
+    fn check_data(data: u128, size: u8) {
+        assert_eq!(truncate(data, Size::from_bytes(size as u64)), data,
```
Comment (oli-obk) [inline src/librustc/mir/interpret/value.rs]: If perf still acts up, this can also be a debug_assert. Users can't act on it anyway and if we get weird crashes, in a dev build we'll immediately see it.

**Review by oli-obk — APPROVED** [review]

File: src/librustc/mir/interpret/value.rs (hunk 4 above)
Comment (RalfJung) [reply src/librustc/mir/interpret/value.rs]: I'd feel better if we could have assertions in release builds at least for either insertion or extraction... but I agree.

**Comment by RalfJung** [comment]
@rust-build timer 2eefc6cff62a771ee8f2ef75a19682171e373e7f

File: src/librustc/mir/interpret/pointer.rs (hunk 1 above)
Comment (Centril) [reply src/librustc/mir/interpret/pointer.rs]: Oops sorry! Missed the `&self` here :) but now you know `'_` can be used at least.

**Comment by RalfJung** [comment]
That test failure definitely looks related...
```
[01:14:26] error[E0379]: trait fns cannot be declared const
[01:14:26]    |
[01:14:26]    |
[01:14:26] LL |     const fn const_val<T: Sized>() -> usize {
[01:14:26]    |     ^^^^^ trait fns cannot be const
[01:14:26] error[E0019]: constant contains unimplemented expression type
[01:14:26]   --> /checkout/src/test/ui/issues/issue-54954.rs:3:24
[01:14:26]    |
[01:14:26]    |
[01:14:26] LL | const ARR_LEN: usize = Tt::const_val::<[i8; 123]>();
[01:14:26] 
[01:14:26] thread 'rustc' panicked at 'attempt to shift left with overflow', src/librustc/mir/interpret/mod.rs:436:5
```
... but I cannot reproduce it locally.

**Comment by RalfJung** [comment]
That's strange, with fewer debug assertions it got *slower*...

Let's try only asserting when constructing, not when ellimninating.

**Comment by RalfJung** [comment]
@rust-timer build f3c792498822725c76f5b5c5febab2a469034195

**Comment by RalfJung** [comment]
This looks good perf-wise!

**Comment by RalfJung** [comment]
This is on the critical path to make Miri work again.

@bors p=1
